File: 3d/src/main/java/io/github/xerooup/frame3d/graphics/mesh/PyramidMesh.java
package io.github.xerooup.frame3d.graphics.mesh;

import io.github.xerooup.frame3d.graphics.Color;

public class PyramidMesh extends BaseMesh {
    private final float[] vertices;
    private final int[] indices;

    public PyramidMesh() {
        this(Color.WHITE);
    }

    public PyramidMesh(Color color) {
        this(1.0f, 1.0f, color);
    }

    public PyramidMesh(float baseSize, float height, Color color) {
        float halfBase = baseSize / 2;
        float halfHeight = height / 2;

        vertices = new float[] {
                // position (x, y, z) + color (r, g, b) + normal (nx, ny, nz) + texCoord (u, v)

                // Top vertex (apex)
                0.0f, halfHeight, 0.0f,  color.r, color.g, color.b,  0, 1, 0,  0.5f, 0.5f,

                // Base vertices
                -halfBase, -halfHeight, halfBase,  color.r, color.g, color.b,  0, -1, 0,  0, 0,
                halfBase, -halfHeight, halfBase,  color.r, color.g, color.b,  0, -1, 0,  1, 0,
                halfBase, -halfHeight, -halfBase,  color.r, color.g, color.b,  0, -1, 0,  1, 1,
                -halfBase, -halfHeight, -halfBase,  color.r, color.g, color.b,  0, -1, 0,  0, 1,

                // Side faces with proper normals and tex coords
                // Front face
                0.0f, halfHeight, 0.0f,  color.r, color.g, color.b,  0, 0.6f, 0.8f,  0.5f, 1.0f,
                -halfBase, -halfHeight, halfBase,  color.r, color.g, color.b,  0, 0.6f, 0.8f,  0.0f, 0.0f,
                halfBase, -halfHeight, halfBase,  color.r, color.g, color.b,  0, 0.6f, 0.8f,  1.0f, 0.0f,

                // Right face
                0.0f, halfHeight, 0.0f,  color.r, color.g, color.b,  0.8f, 0.6f, 0,  0.5f, 1.0f,
                halfBase, -halfHeight, halfBase,  color.r, color.g, color.b,  0.8f, 0.6f, 0,  0.0f, 0.0f,
                halfBase, -halfHeight, -halfBase,  color.r, color.g, color.b,  0.8f, 0.6f, 0,  1.0f, 0.0f,

                // Back face
                0.0f, halfHeight, 0.0f,  color.r, color.g, color.b,  0, 0.6f, -0.8f,  0.5f, 1.0f,
                halfBase, -halfHeight, -halfBase,  color.r, color.g, color.b,  0, 0.6f, -0.8f,  0.0f, 0.0f,
                -halfBase, -halfHeight, -halfBase,  color.r, color.g, color.b,  0, 0.6f, -0.8f,  1.0f, 0.0f,

                // Left face
                0.0f, halfHeight, 0.0f,  color.r, color.g, color.b,  -0.8f, 0.6f, 0,  0.5f, 1.0f,
                -halfBase, -halfHeight, -halfBase,  color.r, color.g, color.b,  -0.8f, 0.6f, 0,  0.0f, 0.0f,
                -halfBase, -halfHeight, halfBase,  color.r, color.g, color.b,  -0.8f, 0.6f, 0,  1.0f, 0.0f
        };

        indices = new int[] {
                // Base
                1, 2, 3,
                3, 4, 1,

                // Faces
                5, 6, 7,
                8, 9, 10,
                11, 12, 13,
                14, 15, 16
        };

        setup(vertices, indices);
    }

    @Override
    public float[] getVertices() {
        return vertices;
    }

    @Override
    public int[] getIndices() {
        return indices;
    }
}

File: 3d/src/main/java/io/github/xerooup/frame3d/graphics/mesh/PyramidMesh.kt
package io.github.xerooup.frame3d.graphics.mesh

import io.github.xerooup.frame3d.graphics.Color

class PyramidMesh(
    private val baseSize: Float = 1.0f,
    private val height: Float = 1.0f,
    color: Color = Color.WHITE
) : BaseMesh() {

    constructor() : this(1.0f, 1.0f, Color.WHITE)
    constructor(color: Color) : this(1.0f, 1.0f, color)

    private val vertices: FloatArray
    private val indices: IntArray

    init {
        val halfBase = baseSize / 2
        val halfHeight = height / 2
        val r = color.r
        val g = color.g
        val b = color.b

        vertices = floatArrayOf(
            // Top vertex (apex)
            0.0f, halfHeight, 0.0f,  r, g, b,  0f, 1f, 0f,  0.5f, 0.5f,

            // Base vertices
            -halfBase, -halfHeight, halfBase,  r, g, b,  0f, -1f, 0f,  0f, 0f,
            halfBase, -halfHeight, halfBase,  r, g, b,  0f, -1f, 0f,  1f, 0f,
            halfBase, -halfHeight, -halfBase,  r, g, b,  0f, -1f, 0f,  1f, 1f,
            -halfBase, -halfHeight, -halfBase,  r, g, b,  0f, -1f, 0f,  0f, 1f,

            // Side faces with proper normals
            // Front face
            0.0f, halfHeight, 0.0f,  r, g, b,  0f, 0.6f, 0.8f,  0.5f, 1.0f,
            -halfBase, -halfHeight, halfBase,  r, g, b,  0f, 0.6f, 0.8f,  0.0f, 0.0f,
            halfBase, -halfHeight, halfBase,  r, g, b,  0f, 0.6f, 0.8f,  1.0f, 0.0f,

            // Right face
            0.0f, halfHeight, 0.0f,  r, g, b,  0.8f, 0.6f, 0f,  0.5f, 1.0f,
            halfBase, -halfHeight, halfBase,  r, g, b,  0.8f, 0.6f, 0f,  0.0f, 0.0f,
            halfBase, -halfHeight, -halfBase,  r, g, b,  0.8f, 0.6f, 0f,  1.0f, 0.0f,

            // Back face
            0.0f, halfHeight, 0.0f,  r, g, b,  0f, 0.6f, -0.8f,  0.5f, 1.0f,
            halfBase, -halfHeight, -halfBase,  r, g, b,  0f, 0.6f, -0.8f,  0.0f, 0.0f,
            -halfBase, -halfHeight, -halfBase,  r, g, b,  0f, 0.6f, -0.8f,  1.0f, 0.0f,

            // Left face
            0.0f, halfHeight, 0.0f,  r, g, b,  -0.8f, 0.6f, 0f,  0.5f, 1.0f,
            -halfBase, -halfHeight, -halfBase,  r, g, b,  -0.8f, 0.6f, 0f,  0.0f, 0.0f,
            -halfBase, -halfHeight, halfBase,  r, g, b,  -0.8f, 0.6f, 0f,  1.0f, 0.0f
        )

        indices = intArrayOf(
            // Base
            1, 2, 3,
            3, 4, 1,

            // Faces
            5, 6, 7,
            8, 9, 10,
            11, 12, 13,
            14, 15, 16
        )

        setup(vertices, indices)
    }

    override fun getVertices() = vertices
    override fun getIndices() = indices
}